### PR TITLE
Implemented Keyring subclass

### PR DESCRIPTION
**Pull request by jimni1222**
## Proposed changes

This PR introduces implementation of Keyring subclass.

- SingleKeyring
- MultipleKeyring
- RoleBasedKeyring

Those classes implement `AbstractKeyring` class.
And for supporting static methods related with Keyring, `KeyringFactory` class is implemented to attach to `caver.keyring`. (`caver.wallet.keyring.encrypt`/`caver.wallet.keyring.encryptV3` are no more supported as a static method.)

Also `signWithKey` and `signWithKeys` functions are combined to `sign`

- In `keyring.sign`, role parameter is not optional and index is optional. If index is undefined, use all private keys in keyring to sign.
- In `transaction.sign`, index and hasher is optional parameters. If index is undefined, use all private keys in keyring to sign. (Same with `caver.wallet.sign`)

And `signFeePayerWithKey` and `signFeePayerWithKeys` functions are combined to `signAsFeePayer`. The hanlding logic for optional parameter will be same with above `sign`

The return type of `signMessage` is changed to return multiple signatures for same index optional policy with `sign` function.
And also `caver.wallet.keyring.recover` moved to `caver.utils.recover`, and `caver.utils.recover` will not receive MessageSigned type parameter.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/caver-js/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/caver-js)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

refer #249 

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: packages/caver-wallet/src/keyring/singleKeyring.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+    /**
+     * returns keys by role.If the key of the role passed as parameter is empty, the default key is returned.
+     *
+     * @param {number} role A number indicating the role of the key. You can use `caver.wallet.keyring.role`.
+     * @return {Array.<PrivateKey>}
+     */
+    getKeyByRole(role) {
+        if (role === undefined) throw new Error(`role should be defined.`)
+        if (role >= KEY_ROLE.roleLast) throw new Error(`Invalid role number: ${role}`)
```
Comment (sirano11) [inline packages/caver-wallet/src/keyring/singleKeyring.js:130]:
```suggestion
        if (role != KEY_ROLE.roleTransactionKey) throw new Error(`Invalid role number: ${role}`)
```

File: packages/caver-wallet/src/keyring/multipleKeyring.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        }
+        return {
+            messageHash,
+            signatures,
+            message,
+        }
+    }
+
+    /**
+     * returns keys by role.If the key of the role passed as parameter is empty, the default key is returned.
```
Comment (sirano11) [inline packages/caver-wallet/src/keyring/multipleKeyring.js:138]:
```suggestion
     * returns keys by role. If the key of the role passed as parameter is empty, the default key is returned.
```

File: packages/caver-wallet/src/keyring/multipleKeyring.js [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+     *
+     * @param {number} role A number indicating the role of the key. You can use `caver.wallet.keyring.role`.
+     * @return {Array.<PrivateKey>}
+     */
+    getKeyByRole(role) {
+        if (role === undefined) throw new Error(`role should be defined.`)
+        if (role >= KEY_ROLE.roleLast) throw new Error(`Invalid role number: ${role}`)
```
Comment (sirano11) [inline packages/caver-wallet/src/keyring/multipleKeyring.js:145]: Don't we need to check negative indexes?

Comment (jimni1222) [reply packages/caver-wallet/src/keyring/multipleKeyring.js:145]: I added "role < 0" check in if condition :) Thank you!

File: packages/caver-wallet/src/keyring/singleKeyring.js [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        signatures.push(key.signMessage(messageHash))
+        return {
+            messageHash,
+            signatures,
+            message,
+        }
+    }
+
+    /**
+     * returns keys by role.If the key of the role passed as parameter is empty, the default key is returned.
```
Comment (sirano11) [inline packages/caver-wallet/src/keyring/singleKeyring.js:123]:
```suggestion
     * returns keys by role. If the key of the role passed as parameter is empty, the default key is returned.
```

File: packages/caver-wallet/src/keyring/singleKeyring.js (hunk 1 above)
Comment (sirano11) [inline packages/caver-wallet/src/keyring/singleKeyring.js]: Don't we need to check negative indexes?

Comment (jimni1222) [reply packages/caver-wallet/src/keyring/singleKeyring.js]: I added validation to all getKeyByRole functions :) Can you refersh?

File: packages/caver-wallet/src/keyring/roleBasedKeyring.js [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+const { fillWeightedMultiSigOptionsForRoleBased } = require('../../../caver-account/src/accountKey/accountKeyHelper')
+const { validateForSigning, validateIndexWithKeys, encryptKey, formatEncrypted } = require('./keyringHelper')
+
+/**
+ * representing a Keyring which includes `address` and `private keys` by roles.
+ * @class
+ */
+class RoleBasedKeyring extends AbstractKeyring {
+    /**
+     * creates a keyring.
```
Comment (sirano11) [inline packages/caver-wallet/src/keyring/roleBasedKeyring.js:35]:
```suggestion
     * creates a RoleBasedKeyring.
```

File: packages/caver-wallet/src/keyring/roleBasedKeyring.js [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+     * @param {string} address - The address of keyring.
+     * @param {Array.<Array<string>>|Array.<Array<PrivateKey>>} keys - The keys to use in roleBasedKeyring.
```
Comment (sirano11) [inline packages/caver-wallet/src/keyring/roleBasedKeyring.js:37]:
```suggestion
     * @param {Array.<Array<string>>|Array.<Array<PrivateKey>>} keys - The keys to use in RoleBasedKeyring.
```

File: packages/caver-wallet/src/keyring/singleKeyring.js [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+const Account = require('../../../caver-account')
+const { validateForSigning, validateIndexWithKeys, encryptKey, formatEncrypted } = require('./keyringHelper')
+
+/**
+ * representing a Keyring which includes `address` and a `private key`.
+ * @class
+ */
+class SingleKeyring extends AbstractKeyring {
+    /**
+     * creates a keyring.
```
Comment (sirano11) [inline packages/caver-wallet/src/keyring/singleKeyring.js:32]:
```suggestion
     * creates a SingleKeyring.
```

File: packages/caver-wallet/src/keyring/multipleKeyring.js [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+const { fillWeightedMultiSigOptionsForMultiSig } = require('../../../caver-account/src/accountKey/accountKeyHelper')
+const { validateForSigning, validateIndexWithKeys, encryptKey, formatEncrypted } = require('./keyringHelper')
+
+/**
+ * representing a Keyring which includes `address` and `private keys`.
+ * @class
+ */
+class MultipleKeyring extends AbstractKeyring {
+    /**
+     * creates a keyring.
```
Comment (sirano11) [inline packages/caver-wallet/src/keyring/multipleKeyring.js:35]:
```suggestion
     * creates a MultipleKeyring.
```

**Pull request merged**